libraries/Microsoft.Bot.Builder/IPropertyAccessor.cs

```diff
@@ -0,0 +1,42 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Text;
+using System.Threading.Tasks;
+
+namespace Microsoft.Bot.Builder
+{
+
+    /// <summary>
+    /// This is metadata about the property including policy info
+    /// </summary>
+    public interface IPropertyAccessor
+    {
+        string Name { get; }
+    }
+
+    /// <summary>
+    /// Interface which defines methods for how you can get data from the context
+    /// </summary>
+    /// <typeparam name="T"></typeparam>
+    public interface IPropertyAccessor<T> : IPropertyAccessor
+    {
+        /// <summary>
+        /// Get the property from the context
+        /// </summary>
+        /// <param name="turnContext"></param>
+        /// <returns></returns>
+        Task<T> GetAsync(ITurnContext turnContext);
```

Comment:
I'm not seeing any value in these methods being async right now. State is always fully loaded at the point where this API kicks in and all of the implementation today is synchronous. The cost of the `Task` allocations and the complexity introduced by making these APIs async doesn't seem worth it... unless there is a plan to do the kind of deferred state loading that I was proposing in my DCR somewhere further down the line?

Reply:
Yeah, that was the idea...

---
In reply to: [203181126](https://github.com/Microsoft/botbuilder-dotnet/pull/760#discussion_r203181126) [](ancestors = 203181126)